controller/volume_controller.go

```diff
@@ -2999,6 +3000,92 @@ func (vc *VolumeController) ResolveRefAndEnqueue(namespace string, ref *metav1.O
 	vc.enqueueVolume(volume)
 }
 
+// createRecurringJobNotExistAndLabel add recurring jobs and group back to a restoring volume and create recurring jobs if not exist.
+func createRecurringJobNotExistAndLabel(log *logrus.Entry, ds *datastore.DataStore, v *longhorn.Volume, jobName string, job *longhorn.RecordVolumeRecurringJob) error {
+	if existJob, err := ds.GetRecurringJob(jobName); err != nil {
+		if !apierrors.IsNotFound(err) {
+			log.WithError(err).Warnf("Failed to get the recurring job %v", jobName)
+			return err
+		}
+		// create the recurring job if not exist
+		log.Debugf("Create the recurring job %v when restoring recurring jobs", jobName)
+		if _, err := ds.CreateRecurringJob(&longhorn.RecurringJob{ObjectMeta: metav1.ObjectMeta{Name: jobName}, Spec: job.JobSpec}); err != nil {
+			log.WithError(err).Warnf("Failed to create a recurring job %v", jobName)
+			return err
+		}
+	} else {
+		// create the recurring job if job name is used and configuration is different.
+		if !reflect.DeepEqual(existJob.Spec, job.JobSpec) {
+			newJobName := util.RandomID()
+			if _, err := ds.GetRecurringJob(newJobName); err != nil {
+				if apierrors.IsNotFound(err) {
+					newJobSpec := job.JobSpec.DeepCopy()
+					newJobSpec.Name = newJobName
+					log.Debugf("Re-create the recurring job %v instead of a existing job: %v when restoring recurring jobs", newJobName, jobName)
+					if _, err := ds.CreateRecurringJob(&longhorn.RecurringJob{ObjectMeta: metav1.ObjectMeta{Name: newJobName}, Spec: *newJobSpec}); err != nil {
+						log.WithError(err).Warnf("Failed to create a recurring job %v", jobName)
+						return err
+					}
+					jobName = newJobName
+				}
+			}
+		}
+	}
+
+	// set Volume recurring jobs/groups
+	if job.FromJob {
+		labelKey := types.GetRecurringJobLabelKey(types.LonghornLabelRecurringJob, jobName)
+		// disabled recurring jobs/group and it would be enabled after the backup restoration is completed.
+		v.Labels[labelKey] = types.LonghornLabelValueEnabled
+	}
+	for _, groupName := range job.FromGroup {
+		groupLabelKey := types.GetRecurringJobLabelKey(types.LonghornLabelRecurringJobGroup, groupName)
+		// disabled recurring jobs/group and it would be enabled after the backup restoration is completed.
+		v.Labels[groupLabelKey] = types.LonghornLabelValueEnabled
+	}
+
+	return nil
+}
+
+//restoreVolumeRecurringJobs create recurring jobs/groups from the backup volume if needed when restoring a backup except DR volume.
+func (vc *VolumeController) restoreVolumeRecurringJobs(v *longhorn.Volume) error {
+	log := getLoggerForVolume(vc.logger, v)
+
+	// Avoid recurring job restoration overrides the new recurring jobs added by users.
+	existingJobs := datastore.MarshalLabelToVolumeRecurringJob(v.Labels)
+	for jobName, job := range existingJobs {
+		if !job.IsGroup || jobName != longhorn.RecurringJobGroupDefault {
+			log.Debugf("Restoring volume recurring jobs had done")
+			return nil
+		}
+	}
+
+	if !v.Spec.Standby {
```

Comment:
NIT: You can check this and directly return at the beginning of this function.

Reply:
Modified. Thanks